shim/src/evaluators.js

```diff
@@ -4,70 +4,67 @@
 import {
   apply,
   arrayJoin,
+  arrayFilter,
   arrayPop,
-  arrayPush,
   create,
   defineProperty,
   getOwnPropertyDescriptors,
   getOwnPropertyNames,
   getPrototypeOf,
   regexpMatch,
-  arrayForEach,
   setPrototypeOf,
   stringIncludes
 } from './commons';
 import { ScopeHandler } from './scopeHandler';
 import { rejectImportExpressions } from './block-imports';
 import { assert, throwTantrum } from './utilities';
 
+// admit many (but not all) legal variable names: starts with letter/_/$,
+// continues with letter/digit/_/$ . It will reject many legal names that
+// involve unicode characters. We use 'apply' rather than /../.match() in
+// case RegExp has been poisoned.
 const identifierPattern = /^[a-zA-Z_$][\w_$]*$/;
 
 // todo: think about how this interacts with endowments, check for conflicts
 // between the names being optimized and the ones added by endowments
 
 function getOptimizableGlobals(safeGlobal) {
-  const constants = [];
   const descs = getOwnPropertyDescriptors(safeGlobal);
 
-  arrayForEach(getOwnPropertyNames(descs), name => {
-    const desc = descs[name];
-    if (typeof name !== 'string') return; // ignore Symbols
-
-    // admit many (but not all) legal variable names: starts with letter/_/$,
-    // continues with letter/digit/_/$ . It will reject many legal names that
-    // involve unicode characters. We use 'apply' rather than /../.match() in
-    // case RegExp has been poisoned.
-
-    if (!regexpMatch(identifierPattern, name)) return;
-
-    // todo: reject keywords, which pass the isIdentifier check, to block
-    // injection attacks. test should use a property name that is itself a
-    // full program
-
-    // getters will not have .writable, don't let the falsyness of
-    // 'undefined' trick us: test with === false, not ! . However descriptors
-    // inherit from the (potentially poisoned) global object, so we might see
-    // extra properties which weren't really there. Accessor properties have
-    // 'get/set/enumerable/configurable', while data properties have
-    // 'value/writable/enumerable/configurable'.
-
-    if (desc.configurable !== false) return;
-    if (desc.writable !== false) return;
-
-    // Check for accessor properties: we don't want to optimize these,
-    // they're obviously non-constant. Setter-only accessors will still have
-    // a 'get' property, but it will be 'undefined', so we only have to test
-    // for 'get', not 'set'
-    if ('get' in desc) return;
-    if ('set' in desc) return;
+  const constants = arrayFilter(getOwnPropertyNames(descs), name => {
+    //
+    // getOwnPropertyNames does ignore Symbols so we don't need this extra check:
+    // typeof name === 'string' &&
+    if (regexpMatch(identifierPattern, name)) {
+      // Optimization: retrieve the descriptor only if we have a valid identifier.
+      const desc = descs[name];
+
+      return (
```

Comment:
I'm using code complexity metrics and this whole function was flagged has having too many return statements. Looking into it, there were several opportunities for optimization.

Reply:
Huh.. if it's faster with the new control flow, that sounds like a good reason to use it, but the new one is less obvious to me: it's combining an if/then and a short-circuiting `&&`. Would it be equally as fast to replace the inner `return(x&&y&&z)` with an `if (x&&y&&z) return true`, and otherwise fall through to the final `return false` outside the `if (regexpMatch)` block?

Reply:
Sorry, I am not following that.

Reply:
I guess I was thinknig of something like this:
```
const constants = arrayFilter(getOwnPropertyNames(descs), name => {
  if (regexpMatch(identifierPattern, name)) {
    const desc = descs[name];
    if (desc.configurable === true &&
        desc.writable === true &&
        'value' in desc) {
      return true;
    }
  }
  return false;
});
```